### PR TITLE
Fix ListenAndServe certificate expiration by preloading certs

### PR DESCRIPTION
**Pull request by brandond**
This change fixes an issue with how `server.ListenAndServe` loads certificates. It sets up a memory -> kubernetes -> memory Secret storage stack, which combined with lazy (on-demand) loading of certificate data, caused the certificate to not be properly renewed when it expired.

It also resolves a regression from #38 - while IPv6 addresses are no longer rejected by the CN regex, they cannot actually be stored in the CN annotation, as semicolons are not allowed in annotation keys. This also fixes handling of hostnames that are more than 41 characters long.

### CODE REVIEW

File: listener.go [hunk 1]
```diff
@@ -405,7 +418,9 @@ func (l *listener) cacheHandler() http.Handler {
 				}
 			}
 
-			l.updateCert(h)
+			if err := l.updateCert(h); err != nil {
+				logrus.Errorf("failed to update cert with HTTP request Host header: %v", err)
+			}
```
Comment (nickgerace) [inline listener.go:423]: Another good catch

File: storage/kubernetes/controller.go [hunk 2]
```diff
@@ -89,18 +89,31 @@ func (s *storage) init(secrets v1controller.SecretController) {
...
+				Name:        s.name,
+				Namespace:   s.namespace,
+				Annotations: secret.Annotations,
+			},
+			Type: v1.SecretTypeTLS,
+			Data: secret.Data,
+		})
+		if err != nil && !errors.IsAlreadyExists(err) {
+			logrus.Warnf("Failed to create Kubernetes secret: %v", err)
+		}
```
Comment (nickgerace) [inline storage/kubernetes/controller.go:106]: Should we add retry logic here or will the warning log suffice? I'm led the believe that a non-"IsNotFound" error will likely reappear upon retry, so I'm good with this "as-is", but thought I'd ask if you considered it.

**Review by nickgerace — APPROVED** [review]
Thank you! The design lines up with what we discussed, so this LGTM.

File: storage/kubernetes/controller.go (hunk 2 above)
Comment (brandond) [reply storage/kubernetes/controller.go:106]: The Kubernetes Secret also gets watched in the background, and the sync from the watch plus the write-back of any local updates to the secret will ensure that it is eventually consistent.

**Pull request merged**